### PR TITLE
New merging process

### PR DESCRIPTION
**Pull request by maximedenes**
The merging process for the last few release cycles has been very centralized. While this was effective at stabilizing our processes, it puts a strain on me and leads to issues like #5986.

I won't volunteer for managing the 8.9 merges, so here is a proposal to distribute the process instead. This is reasonably urgent matters, since we shouldn't block PRs for `master`, now that 8.8 has branched off.

I created a `CODEOWNERS` file, inspired from the Maintainers page of the wiki. Please help me fill the gaps. As described in the notes, each entry needs two GitHub nicknames registered as maintainers (owners).

### CODE REVIEW

**Comment by herbelin** [comment]
I'm supporting the principle of a distributed merging process.

**Comment by herbelin** [comment]
Whoever (a singular or plural "who") merge(s), it would help (at least me) that the reviewing/merging process is made more codified. How to ensure that enough time is given so that the discussions which have to happen can happen? How to reasonably ensure that a PR does not stall pending on someone's action which does not come?

At some time, I somehow had in mind adopting a "canonical" distributed merging process, namely that core developers merge their PRs themselves, after an agreement is reached following a reviewing/merging protocol adopted by everyone, but deciding who merges by component is also a good idea. In any case, it is clear that for contributions from non-core developers, merging has to be done by a core developer.

For the period preceding releases, we could switch back to a more centralized mode. A deadline for submission of PRs for this release 3 weeks in advance (since 1 week was too short for 8.7 and 8.8). During the first 2 weeks of these 3 weeks, a reduced number of persons (one person with enough time available or two persons working in close interaction) coordinate the convergence process, typically the way @maximedenes or @zimmi48 were doing it, with the objective that 1 week before the release, only minor fine-tuning, only last minute bug fixes, only last minute infrastructure problems are treated (in particular, updating credits should also be ready to be reviewed one week in advance, for a merge say 3 days before building packages). During the convergence period, merging is made only by those people who manage the release process and who have a global overview of the ongoing schedule. Everyone is however supposed to be reactive for commenting, reviewing, amending the PRs being in the pipeline.

**Comment by ejgallego** [comment]
This is a great idea thanks @maximedenes even if I still feel that some global changes should be taken care of by a "RM" manager that of course can be a rotating position.

I will try to contribute more thoughts soon, but for now let me state a wish before we move in this direction, which is that we try to improve the CI system a bit.

I think that much of the strain you suffered as a "merge-master" [apart from us whiners] was due to slow response of our CI infrastructure, so IMHO it is critical we improve this so codeowners don't merge broken stuff.

By the way thank you very much for all the work energy and service you've done in the very hard and not very well gratified position as a "merge-master",

**Comment by gares** [comment]
> By the way thank you very much for all the work energy and service you've done in the very hard and not very well gratified position as a "merge-master",

I join and say "thanks for your patience".

File: .github/CODEOWNERS
Comment (gares) [inline .github/CODEOWNERS]: `toplevel` contains a lot of boring code everybody understands, + a few tricky bits about initialization that I may know better.  So I'm fine having my name here.

`vernac` I really don't know why my name is in there:
```
vernac/assumptions.mli         vernac/declareDef.mli   vernac/proof_using.mli
vernac/auto_ind_decl.mli       vernac/explainErr.mli   vernac/record.mli
vernac/classes.mli             vernac/himsg.mli        vernac/search.mli
vernac/class.mli               vernac/indschemes.mli   vernac/topfmt.mli
vernac/comAssumption.mli       vernac/lemmas.mli       vernac/vernacentries.mli
vernac/comDefinition.mli       vernac/locality.mli     vernac/vernacinterp.mli
vernac/comFixpoint.mli         vernac/metasyntax.mli   vernac/vernacprop.mli
vernac/comInductive.mli        vernac/mltop.mli        vernac/vernacstate.mli
vernac/comProgramFixpoint.mli  vernac/obligations.mli
```
The only (trivial) file in there I wrote is `proof_using`, and the only code I've touched is the dispatch of the interpreter. All the rest is the implementation of vernacular commands. I'm not comfortable with my name being here. For sure it lacks @mattam82 for program/obligations. It should also contain the code of Arguments, that @maximedenes completely rewrote, doesn't it?

File: .github/CODEOWNERS
Comment (gares) [inline .github/CODEOWNERS]: This looks stuff for @herbelin (notations, intern/exern-alization, ...)

File: dev/doc/MERGING.md
Comment (gares) [inline dev/doc/MERGING.md]: "should" is often misunderstood by non native speakers.
Should we conform to: https://www.ietf.org/rfc/rfc2119.txt ?

**Comment by herbelin** [comment]
And as far as I'm concerned, it is not because there are things which I would need to be improved that what works well should not be made explicit too. And to partially repeat what I said in #6954, thanks also for your patience, energy, reliability.

File: dev/doc/MERGING.md
Comment (gares) [reply dev/doc/MERGING.md]: Eg, here I would use MUST NOT, the only exception I can think of is during the actual release, but in that case it is the release manager that may merge his own PR (to, eg, fix the version number...)

**Comment by herbelin** [comment]
To the CI experts: I know this has already been discussed, but could we have a way to deactivate the running of CI, with an explicit field in the template asking whether a test is required for the PR or not (the current PR being a typical example of useless - if not damageable - dissipation of CI power).

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS]: Wrong nickname. You meant `@gares`.

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS]: That's `@forestjulien`.

**Comment by SkySkimmer** [comment]
>To the CI experts: I know this has already been discussed, but could we have a way to deactivate the running of CI, with an explicit field in the template asking whether a test is required for the PR or not (the current PR being a typical example of useless - if not damageable - dissipation of CI power).

You can put `[skip ci]` in a commit message but I don't know if that works for PRs.

**Comment by Zimmi48** [comment]
I can only support this approach because it is one that I wanted to advocate.

> How to ensure that enough time is given so that the discussions which have to happen can happen? How to reasonably ensure that a PR does not stall pending on someone's action which does not come?

As you said in an e-mail, bots could help with this and I'm willing to try out some bot programming.

> For the period preceding releases, we could switch back to a more centralized mode.

:+1: (that's the only thing that was preventing me to ask a more distributed management)

> I think that much of the strain you suffered as a "merge-master" [apart from us whiners] was due to slow response of our CI infrastructure, so IMHO it is critical we improve this so codeowners don't merge broken stuff.

:+1: 

> the only exception I can think of is during the actual release, but in that case it is the release manager that may merge his own PR

:+1:

File: .github/CODEOWNERS
Comment (maximedenes) [reply .github/CODEOWNERS]: And what about the secondary maintainer of that part?

**Comment by maximedenes** [comment]
Updated, thanks for the feedback. I would like to put this in place ASAP, but with the current list of maintainers it means we are going to get stuck fairly quickly. So I'll suggest new maintainers (especially secondary ones), please let me know if it is not ok to you when I put your name.

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS]: I suggest @JasonGross for `Compat/`

**Comment by maximedenes** [comment]
@ejgallego I would like to add you as secondary maintainer of `clib`, `interp`.

@herbelin I would like to add you as secondary maintainer of `btauto`, `ltac`, `rtauto`.

@ppedrot I would like to add you as secondary maintainer of `nsatz`.

@silene I would like to add you as secondary maintainer of `coq_makefile`.

`omega` and `romega` also need a secondary maintainer, please suggest names!

A few more parts are missing maintainers, but hopefully it won't block the merge process. If it does, we can call for more volunteers.

File: .github/CODEOWNERS
Comment (maximedenes) [reply .github/CODEOWNERS]: Sure, if @JasonGross is ok. Note however that since he's the main contributor to those files, we'd need a secondary maintainer.

**Comment by silene** [comment]
> I would like to add you as secondary maintainer of `coq_makefile`.

Fine with me. You can also add me to `kernel/byterun`.

Do I understand correctly that `CODEOWNERS` is a file natively handled by github and that people mentioned in it will be automatically pinged on pull requests?

**Comment by maximedenes** [comment]
> Do I understand correctly that CODEOWNERS is a file natively handled by github and that people mentioned in it will be automatically pinged on pull requests?

Yes, that's correct. They will be assigned as reviewers with a special "code owner" status and a nice icon.

**Comment by herbelin** [comment]
My understanding of the rules of the game is that when assigned by the robot, it becomes our responsibility to organize the discussion on the PR, to do a review and/or to sollicit reviews from others, to ensure that the PR does not stall, to check that relevant doc and tests have been done, to eventually run the merge script when the PR seems discussed and supported enough and that ci is green? If this view is shared, I can propose a clarification in `MERGING.md` file to reflect it.

@skyskimmer: Thanks, so I shall try to follow how the support for this "skip" evolves.

**Comment by maximedenes** [comment]
> My understanding of the rules of the game is that when assigned by the robot

Note that in most cases, the robot will assign many reviewers. One of them will become the "assignee" following the process described in the document.

> it becomes our responsibility to organize the discussion on the PR

I don't know, it may also be the responsibility of the author (or shared).

> to do a review and/or to sollicit reviews from others

Yes, that's already written.

> to ensure that the PR does not stall

How do you plan to ensure that? As far as I'm concerned, I would strongly object to committing to something I don't have control over.

> to check that relevant doc and tests have been done

Yes, that's already written.

> to eventually run the merge script when the PR seems discussed and supported enough and that ci is green

Yes, that's already written. Maybe I should add that if any core developer requests more discussion, it shouldn't be merged until that happened (it is the current process).

**Comment by herbelin** [comment]
> > it becomes our responsibility to organize the discussion on the PR
> 
> I don't know, it may also be the responsibility of the author (or shared).

Yes, it can also be the responsibility of the author, or a shared responsibility, at least when it is about core developers who are used to work together. So, indeed, it can be one or the other.

It remains the case of an external contributor: it is maybe asking a lot that he/she organizes the discussion on his/her own PR, in the sense that he/she does not necessarily know our habits, and does not necessarily feel legitimate to take a coordination role as an external person.

>> to ensure that the PR does not stall
> 
> How do you plan to ensure that? As far as I'm concerned, I would strongly object to committing to something I don't have control over.

I think this corresponds to your "regularly ping the author (if changes should be implemented) or the reviewers (if reviews are missing)". Or, in case the author takes the responsibility of the coordination of the discussion "regularly ping the reviewers (if reviews are missing) or the commenters (if answers to questions are missing)".

In practice, I easily feel uncomfortable with re-pinging. How to evaluate the reasonable delay before pinging again without being intrusive or impolite? A possibility is that when asking the author about changes or the reviewers about a review, one asks them at the same time to propose an "estimated time of arrival". Then "ensuring that the PR does not stall" shall mean "ensuring that the author or reviewer did the changes or review in the time he promised". It is trivial matter, but I feel that codifying this shall avoid misunderstandings, as we have seen, of waiting for someone and nothing coming, and a PR missing a release. Then, the responsibility of the assignee is that noone procrastinates more than reasonable (but, of course, the assignee can not be considered responsible in place of someone else not satisfying a promise in spite of the efforts of the assignee).

> > to eventually run the merge script when the PR seems discussed and supported enough and that ci is green

> Yes, that's already written. Maybe I should add that if any core developer requests more discussion, it shouldn't be merged until that happened (it is the current process).

Yes, that's my point, and to avoid misunderstandings as we have seen, time should be explicitly given so that "any core developer who wants to request more discussion has the opportunity to do so". So, typically, one could change the sentence "Once all reviewers approved the PR, the assignee is expected to follow the merging process described below" into "The assignee ensures that any requests for more discussion have been granted. When the discussion has converged and all reviewers have approved the PR, the assignee is expected to follow the merging process described below."

**Comment by gares** [comment]
> How to evaluate the reasonable delay before pinging again without being intrusive or impolite?

I think that we should codify the crucial technical steps we should follow and keep the document short, like a checklist.

For all the rest, I believe common sense is just ok.

**Comment by maximedenes** [comment]
@herbelin are you ok with the parts of the system for which I mentioned your name as a secondary maintainer above?

**Comment by herbelin** [comment]
@maximedenes: You can put my name wherever you want (including in `Vectors`). But about ensuring that discussion is granted whenever needed, I think it is important to mention it. My analysis is that it has been one of the causes of frustrations and misunderstandings in the past.

File: .github/CODEOWNERS
Comment (JasonGross) [reply .github/CODEOWNERS]: Sure, though I don't think I have push bits for running the merge script, which according to `dev/doc/MERGING.md` would be my responsibility for changes primarily about `Compat` files?

**Comment by maximedenes** [comment]
> So, typically, one could change the sentence "Once all reviewers approved the PR, the assignee is expected to follow the merging process described below" into "The assignee ensures that any requests for more discussion have been granted. When the discussion has converged and all reviewers have approved the PR, the assignee is expected to follow the merging process described below."

Done.

**Comment by ejgallego** [comment]
> To the CI experts: I know this has already been discussed, but could we have a way to deactivate the running of CI, with an explicit field in the template asking whether a test is required for the PR or not (the current PR being a typical example of useless - if not damageable - dissipation of CI power).

Disabling CI precisely goes against the spirit of the CI philosophy, which is to have a continuous deployment. I would only support such a feature if implemented in a consistent way, as for example as a precondition on the CI scripts themselves.

Improving the energy efficiency of the release process is of course welcome [even if more energy would be saved by not using consuming disposable cups in the meetings], but in practical terms the best way to save computing energy is by moving the CI system to use _residual_ computing power available on the cloud.

**Comment by ejgallego** [comment]
> @ejgallego I would like to add you as secondary maintainer of clib, interp.

Ok.

**Comment by ejgallego** [comment]
> For all the rest, I believe common sense is just ok.

I'm with Enrico here, I think that too much codification just obscures what should be a matter of dynamics and common sense. I cannot but disagree with saying that "too much pinging is problematic", the more we communicate the better. Also I don't think that we've suffered from a real problem where many PRs have been merged before a discussion would happen, but in any case this is a very easy to solve process as reverting is easy and IMHO we shouldn't be afraid of it.

@maximedenes , a bit of a key part that is missing in the owners file is the build / makefile infrastructure, isn't it?

**Comment by maximedenes** [comment]
> @maximedenes , a bit of a key part that is missing in the owners file is the build / makefile infrastructure, isn't it?

I was going to put the first who would mention it. :D 

No seriously, it was an omission, sorry about that. Should I put you and @letouzey, @letouzey and you? Somebody else?

**Comment by ejgallego** [comment]
> No seriously, it was an omission, sorry about that. Should I put you and @letouzey, @letouzey and you? Somebody else?

I am OK helping with that part.

**Comment by maximedenes** [comment]
Ok, this is starting to look good. If no objections, I'll merge tomorrow. We can always update this file. I volunteered to maintain it :)

As @JasonGross mentioned, code owners need to receive push rights. But I don't feel authorized to grant push rights, so let's do that separately.

**Comment by herbelin** [comment]
Are there other opinions than from @Zimmi48 about the proposed protocol for the process of convergence towards the release? (And the more I'm thinking about it, the more I'm wondering whether it should not even be a 1 month period.)

**Comment by ejgallego** [comment]
> Are there other opinions than from @Zimmi48 about the proposed protocol for the process of convergence towards the release? (And the more I'm thinking about it, the more I'm wondering whether it should not even be a 1 month period.)

First, I think we can distinguish two different kind of PRs: straightforward PRs, such as trivial changes, non-complex bugfixes, tweaks to the CI, etc... In this case, it seems to me that the maintainers of each subsystem can just approve and merge quickly.

On the other hand we have "larger-scale" PRs that are likely to involve deeper design decisions and may touch several subsystems. Also, I'd put in that category PRs that even if simple, require a large amount of discussion [for example, the one adding `Export Set`].

So I think we all agree that PRs in the first case can be just merged without too much of a delay, on the other, PRs belonging to the second category may require a couple of weeks of discussion, tweaks, more careful review, etc... and a more coordinated merge due to the likely interactions with other PRs.

Also, it seems that "continous" evaluation of these second class of PRs is a bit tiring.

I liked @Zimmi48 's idea of having a couple of `alpha` checkpoints plus the `beta` one. So indeed, we could say that such checkpoints happen once every 5 weeks. Then, the tree would open for such merges for 4 weeks, then the RM would take over and evaluate the state of PRs of the second category, merge the ones that are ready, and ask the authors of PRs needing more work to reschedule based on the perceived effort to submit a revision.

**Comment by ejgallego** [comment]
Another very important point seems to be packaging. We have two _custom_ :baby_chick: packaging systems **in-tree** in order to generate OSX and Windows binaries. Moreover packages are tied to releases [which seems pretty constraining] For example, how are you gonna rotate the RM if the chosen person has no access to a windows / OSX machines. That alone seems to severely limit the number of available candidates.

I think that we should: a) decouple packaging from the core codebase [I mean moving all specific code to `dev/pkg/opam`, `dev/pkg/windows`, etc..., b) look for owners for each of the packages.

An interesting thing more "modern" build systems do is that they eliminate the local/global distinction, in the sense that for example Dune will create a hierarchy under `_build` [or the specified dir] that is exactly what you get after install.

In fact, Dune for example can work with cross compilation so it wouldn't be too crazy to generate windows packages this way. For sure, windows or mac packages can seed the build system using their own workspace.

**Comment by Zimmi48** [comment]
Let's not hijack this PR, which is only about the merging process, to talk about the release process please. The more we separate different issues, the more efficient it will be.

On the topic of "is it the responsibility of the assignee to make sure the PR doesn't stale", I think this should be indeed the responsibility of the assignee to know who is expected to do something for a PR to progress, and with which ETA. In case of stalling because of unresponsiveness of the author, or because there is obviously no consensus to be found between reviewers, then it should also be the responsibility of the author to close the PR. We shouldn't keep open stalled PRs (we have many right now). However, this idea could be discussed separately (or even tested before being written down) so I don't think that the discussion on this topic should interfere with the merging of the present PR. The documentation as it is right now LGTM as a first shot.

On the question of giving push access:
- It is easy to create a new sub-team of "contributors" to which we give push access, but only for merging PRs that have been approved by code owners (and never to push commits directly).
- We can restrict this to the master branch only and keep the stable branches the exclusive domain of repo admins.
- We could also require status checks to have passed but given the current rate of failure of our CI, it would be counter-productive. We could restrict this requirement to a subset of the statuses though (for instance, require that at least Travis has passed) and this would not be enforced for repo admins and that would still allow for "emergency merging".
- FTR other things that we can enable are "automatic dismissal of stale approved reviews" (to ensure that the code owner validates the latest version of the PR before merging it) and "restriction of the power to dismiss a review" (I don't think that this would be useful right now). 
- We can also require the different code owners to explicitly consent to the merging documentation before we add them to the pushers team (if we do this, it's better to do it in a separate issue once this PR is merged).

In any case, whatever is decided on this aspect, I can take care of implementing it quickly.

BTW @maximedenes: owners are missing for packaging scripts, `dev/ci` and toplevel documentation files.

**Comment by maximedenes** [comment]
> owners are missing for packaging scripts, dev/ci

Fixed.

> and toplevel documentation files.

Do you mean `README.md` and others?

**Comment by herbelin** [comment]
> First, I think we can distinguish two different kind of PRs: straightforward PRs, such as trivial changes, non-complex bugfixes, tweaks to the CI, etc... In this case, it seems to me that the maintainers of each subsystem can just approve and merge quickly.

This was my position at some time, but I realized that several people, starting with someone called @ejgallego, were interested in reviewing, commenting, even what looked to me like without particular importance. I then started to receive particularly good remarks on my PRs from various persons and to appreciate that time is given for discussion even for PRs which look straightforward for the author. Of course, it is a question of compromise with time, and sometimes there are urgencies, but, considering that with one people merging, merging of a trivial bug fix shall anyway take sometimes more than a week, we can easily allocate 24h to ensure that everyone has told what he had to say. (It would be a physical meeting, 5mns would be enough to look at the assembly and to ask for possible further questions, but since we are online, we cannot expect everyone to be online continuously).

I mean, I'm not opposed per se to more informality, but in case the informality of someone does not match the idea of informality that someone else has, and a disagreement arrives after the fact (as it happened), we also need sometimes to improve our way to communicate... Typically, if a PR is merged while someone has still something to say, how do you expect that he/she tells it: by opening a new complementary PR? by asking for a revert? ...

There are otherwise many interesting points to discuss... but maybe at a later time.

**Comment by Zimmi48** [comment]
> Do you mean README.md and others?

Yep, README, INSTALL, CONTRIBUTING...

**Comment by maximedenes** [comment]
@Zimmi48 Do you volunteer for these files? :innocent:

**Comment by Zimmi48** [comment]
I do.

**Comment by Zimmi48** [comment]
@herbelin 
> This was my position at some time

No, you missed an important difference which is that the author of a PR cannot self-approve a PR so cannot merge it. What your position was at some time was that trivial fixes could be committed directly. What is proposed here is that a PR is always reviewed before merging, by a code owner. If the code owner thinks that the patch is safe, then they may proceed and merge it without any wait: but they are the code owner, they should know what they are doing. Of course, in the case of a larger patch, or a feature, a reasonable delay should be used, but it's not necessarily required to write that down. At least, we can try without writing it down and if it fails, we can update the merging doc.

**Comment by ejgallego** [comment]
@Zimmi48 _

> Let's not hijack this PR, which is only about the merging process, to talk about the release process please. The more we separate different issues, the more efficient it will be.

Oh interesting, I had somehow implicitly linked them. 

@herbelin 
> This was my position at some time, but I realized that several people, starting with someone called @ejgallego, were interested in reviewing, commenting, even what looked to me like without particular importance

Oh sure, I didn't mean quickly == without review. Maybe we have something different in mind when we talk of "straightforward" PRs. For example, I don't consider those adding a feature to be "straightforward" but of course OMMV. I think that Théo summarized it better.

> Typically, if a PR is merged while someone has still something to say, how do you expect that he/she tells it: by opening a new complementary PR? by asking for a revert? ...

The discussion on the merged PR is still open and I think the involved people is still notified. I guess that would depend but it seems easy to do. I would expect most additional commits to be incorporated in a new PR, or else you could open an issue to discuss. Either way, I see post-merge discussion as healthy and necessary.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
Quoting from #7020 where I posted on error:

> It probably applies only on fresh PRs, which make sense because reviewers may have been assigned already otherwise.

Maybe that was discussed by I may have missed it, indeed we need a "grandfathering" strategy for old PRs?

Maybe you could do the initial triage [merging/assigning if not yet ready] ?

**Comment by maximedenes** [comment]
Yes, we discussed it. After seeing that there were 129 open PRs, I suggested we do a collaborative effort :)

**Comment by ejgallego** [comment]
The thing is that the PRs I can triage better are of course my own PRs, but I am not sure how should I proceed here for "self-triaging".

**Comment by maximedenes** [comment]
> The thing is that the PRs I can triage better are of course my own PRs,

Yes, I think it is a good idea.

> but I am not sure how should I proceed here for "self-triaging".

For each path touched by your PR, you'll find the corresponding code owners in `.github/CODEOWNERS`, and you assign them as reviewers.

**Comment by ppedrot** [comment]
@maximedenes OK to be added to `nsatz` owners.

**Pull request merged**
